# Исследование надежности заемщиков


## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd
data = pd.read_csv('/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
# код для построения сводной таблицы:
data_child = data.pivot_table(index=['children'], columns = 'debt', values='education_id', aggfunc='count')
data_child[1] = data_child[1].fillna(0)
data_child = data_child.rename(columns={0:'Нет задол-сти',1:'Есть задол-сть'})
data_child['Всего клиентов'] = data_child['Нет задол-сти']+data_child['Есть задол-сть']
data_child['% должников'] = data_child['Есть задол-сть']/data_child['Всего клиентов']*100

#код для форматирования в таблице знаков после запятой:
data_child [['Нет задол-сти','Есть задол-сть','Всего клиентов']] = data_child [['Нет задол-сти','Есть задол-сть','Всего клиентов']].astype('int')
data_child['% должников'] = data_child['% должников'].round(decimals=1)
print('Клиенты, сгруппированные по количеству детей в семье:')
print(data_child)




Клиенты, сгруппированные по количеству детей в семье:
debt      Нет задол-сти  Есть задол-сть  Всего клиентов  % должников
children                                                            
0                 13028            1063           14091          7.5
1                  4364             444            4808          9.2
2                  1858             194            2052          9.5
3                   303              27             330          8.2
4                    37               4              41          9.8
5                     9               0               9          0.0


**Вывод:** 
Самый низкий риск не возврата кредитов в срок у клиентов без детей (только 7.5% таких клиентов имели задолженность по возврату кредита). Среди клиентов, имеющих одного-двух детей, доля несвоевременно погасивших платеж практически одинаковая и составляет 9.2-9.5%. Категория клиентов с тремя и более детьми малочислена, тем не менее 9.8% клиентов с четырьмя детьми имели задолженность по возврату кредита в срок, это чаще, чем в остальных категориях. Семьи с пятью детьми вернули все кредиты вовремя, но в этой категории всего 9 клентов. Таким образом, бездетные клиенты чаще возвращают кредиты в срок, чем клиенты с детьми. При этом, количество детей координально не влияет на риск не возврата кредита.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
data_family = data.pivot_table(index=['family_status'], columns = 'debt', values='family_status_id', aggfunc='count')
data_family = data_family.rename(columns={0:'Нет задол-сти',1:'Есть задол-сть'})
data_family['Всего кл-в'] = data_family['Нет задол-сти']+data_family['Есть задол-сть']
data_family['% должников'] = data_family['Есть задол-сть']/data_family['Всего кл-в']*100
data_family['% должников'] = data_family['% должников'].round(decimals=1)
print(data_family)

debt                   Нет задол-сти  Есть задол-сть  Всего кл-в  % должников
family_status                                                                
Не женат / не замужем           2523             273        2796          9.8
в разводе                       1105              84        1189          7.1
вдовец / вдова                   888              63         951          6.6
гражданский брак                3749             385        4134          9.3
женат / замужем                11334             927       12261          7.6


**Вывод:** 
Зависимость между семейным положением и возвратом кредита в срок существует. Категория "женат/замужем" является самой крупной, при этом доля клиентов в этой категории, имевших задолженность по возврату кредита в срок, составила 7.6% Также низкие риски не возврата кредитов в срок у клиентов в статусе  вдова/вдовец (6.6%) и в разводе (7.1%), но эти категории не значительны в общей численности клиентов банка. Не женатые (не замужние) клиенты и клиенты, находящиеся в гражданском браке, чаще других имеют проблемы с возвратом кредитов. Таким образом, клиенты, состоящие в официальном браке реже имеют задолженность по возврату кредита в срок, чем клиенты в статусе не женат/не замужем или находящиеся в гражданском браке. 

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [24]:
data_income = data.pivot_table(index=['total_income_category'], columns = 'debt', values='total_income', aggfunc='count')
data_income = data_income.rename(columns={0:'Нет задол-сти',1:'Есть задол-сть'})
data_income['Всего кл-в'] = data_income['Нет задол-сти']+data_income['Есть задол-сть']
data_income['% должников'] = data_income['Есть задол-сть']/data_income['Всего кл-в']*100
data_income['% должников'] = data_income['% должников'].round(decimals=1)
print(data_income)

debt                   Нет задол-сти  Есть задол-сть  Всего кл-в  % должников
total_income_category                                                        
A                                 23               2          25          8.0
B                               4660             354        5014          7.1
C                              14568            1353       15921          8.5
D                                328              21         349          6.0
E                                 20               2          22          9.1


**Вывод:** 
Категория "D" - клиенты с доходом от 30 тыс. до 50 тыс. - чаще других категорий возвращают все кредиты в срок (доля должников среди клиентов с таким доходом - 6%), но при этом количество таких клиентов очень незначительно. Больше всего клиентов в категории "С" (50 тыс.- 200 тыс.) и здесь риск не возврата кредита в срок возрастает - доля должников в этой категории 8.5%. Категория "В" (доход от 200.0 тыс. до 1 млн.)не так многочисленна как "С", но в ней доля клиентов, не вернувших кредит в срок снижается и составляет 7.1%. Самый высокий риск не своевременного возврата кредита у категории "Е" - доход до 30 тыс. руб. (9.1% имели задолженности), однако таких клиентов совсем мало - 22 человека. Таким образом, должниками чаще становятся клиенты с доходом  50-200 тыс., а вот в категории с доходом от 200 тыс. до 1 млн. должников меньше. Число клиентов из других категорий не значительно, тем не менее нельзя не учитывать высокий риск задолженности у клиентов с доходом до 30 тыс.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [25]:
data_purpose = data.pivot_table(index=['purpose_category'], columns = 'debt', values='total_income', aggfunc='count')
data_purpose = data_purpose.rename(columns={0:'Нет задол-сти',1:'Есть задол-сть'})
data_purpose['Всего кл-в'] = data_purpose['Нет задол-сти']+data_purpose['Есть задол-сть']
data_purpose['% долж-в'] = data_purpose['Есть задол-сть']/data_purpose['Всего кл-в']*100

data_purpose['% долж-в'] = data_purpose['% долж-в'].round(decimals=1)
print(data_purpose)


debt                      Нет задол-сти  Есть задол-сть  Всего кл-в  % долж-в
purpose_category                                                             
операции с автомобилем             3879             400        4279       9.3
операции с недвижимостью           9971             780       10751       7.3
получение образования              3619             369        3988       9.3
проведение свадьбы                 2130             183        2313       7.9


**Вывод:** 
Чаще всего клиенты берут кредиты на операции с недвижимостью и именно по таким кредитам самый низкий риск несвоевременных платежей. Доля клиентов, имевших задолженность, в этой группе составила 7.3% Клиенты, взявшие кредит на проведение свадьбы, также стараются совершать все платежи в срок (доля должников в этой категории - 7.9%). Но надо учитывать, что кредит с целью провести свадьбу берут реже всего. Кредиты на получение образования и на операции с автомобилем имеют более высокие риски задержки платежей. 9.3% клиентов таких групп имели задолженность. Таким образом, платежи по кредитам на операции с недвижимостью чаще поступают в срок, чем по кредитам, взятым с другими целями.   

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Пропуски обнаружены в колонках с трудовым стажем и ежемесячным доходом, это два взаимосвязанных показателя. Скорей всего данные от работодателя по этим клиентам еще не переданы в банк или не занесены в базу. Нужно уточнить эту информацию у сотрудников банка. В случае, если все данные занесены в базу, делаем вывод, что это клиенты, которые только что устроились на работу и у них такие данные отсутсвуют. 

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Заполнение пропусков медианным значением дает более объективную картину. Например  ежемесячный доход, соответсвующий категории "С" составляет значительную часть выборки, при этом в таблице есть значения дохода свыше 1-2 млн., их немного, но суммы очень большие и могут исказить данные о среднем доходе клиента, завысив его.  

### Шаг 4: общий вывод.

На основании проведенного иследования можно сделать вывод, что существует зависимость возврата кредита в срок от ряда факторов. Это семейное положение, наличие детей, среднемесячный доход клиента и цель кредита. Чаще всего оплачивают в срок кредиты клиенты в статусе женат/замужем, без детей и с ежемесячным доходом от 200 тыс. до 1 млн., взятые с целью проведения операций с недвижимостью. В группе риска не женатые/не замужние клиенты или клиенты, состоящие в гражданском браке, имеющие детей, ежемесячный доход которых составит от 50 тыс. до 200 тыс. 